From 98b0dc572d503f77786bcd7a39ca38e98cb0f2cc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Vivek=20Das=C2=A0Mohapatra?= <vivek@collabora.com>
Date: Tue, 24 Sep 2024 01:19:59 +0100
Subject: [PATCH] Defer theme loading until we know we need it

This can save a significant amount of time in quiet mode.
Loading the `breeze' theme from a btrfs root takes around
1.4 seconds.

For comparison our total bootloader time (including theme load)
is about 2.5 seconds.
---
 util/grub.d/00_header.in | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/util/grub.d/00_header.in b/util/grub.d/00_header.in
index d26062948..12f9901d0 100644
--- a/util/grub.d/00_header.in
+++ b/util/grub.d/00_header.in
@@ -255,35 +255,38 @@ if [ "x$gfxterm" = x1 ]; then
 
 	prepare_grub_to_access_device `${grub_probe} --target=device "$GRUB_THEME"`
 	cat << EOF
-insmod gfxmenu
+if [ "\$timeout:\$timeout_style" = "-1:menu" ];
+then
+  insmod gfxmenu
 EOF
 	themedir="`dirname "$GRUB_THEME"`"
 	for x in "$themedir"/*.pf2 "$themedir"/f/*.pf2; do
 	    if [ -f "$x" ]; then
 		cat << EOF
-loadfont (\$root)`make_system_path_relative_to_its_root $x`
+  loadfont (\$root)`make_system_path_relative_to_its_root $x`
 EOF
 	    fi
 	done
 	if [ x"`echo "$themedir"/*.jpg`" != x"$themedir/*.jpg" ] || [ x"`echo "$themedir"/*.jpeg`" != x"$themedir/*.jpeg" ]; then
 	    cat << EOF
-insmod jpeg
+  insmod jpeg
 EOF
 	fi
 	if [ x"`echo "$themedir"/*.png`" != x"$themedir/*.png" ]; then
 	    cat << EOF
-insmod png
+  insmod png
 EOF
 	fi
 	if [ x"`echo "$themedir"/*.tga`" != x"$themedir/*.tga" ]; then
 	    cat << EOF
-insmod tga
+  insmod tga
 EOF
 	fi
 	    
 	cat << EOF
-set theme=(\$root)`make_system_path_relative_to_its_root $GRUB_THEME`
-export theme
+  set theme=(\$root)`make_system_path_relative_to_its_root $GRUB_THEME`
+  export theme
+fi
 EOF
     elif [ "x$GRUB_BACKGROUND" != x ] && [ -f "$GRUB_BACKGROUND" ] \
 	    && is_path_readable_by_grub "$GRUB_BACKGROUND"; then
-- 
2.39.2

